File: Leetcode/Python/_106.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        def build(left, right):
            if left > right:
                return None

            val = postorder.pop()
            node = TreeNode(val)

            left_index = mapping_inorder[val]
            node.right = build(left_index + 1, right)
            node.left = build(left, left_index - 1)
            return node


        mapping_inorder = {val: i for i, val in enumerate(inorder)}
        return build(0, len(inorder) - 1)

File: Leetcode/cpp/106.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        int n = inorder.size();
        int index = n - 1;
        std::unordered_map<int, int> umap;
        for (int i = 0; i < n; ++i) {
            umap[inorder[i]] = i;
        }
        std::function<TreeNode*(int, int)> build = [&](int left, int right) -> TreeNode* {
            if (left > right) {
                return nullptr;
            }
            TreeNode* node = new TreeNode(postorder[index]);
            int idx = umap[postorder[index--]];
            node->right = build(idx + 1, right);
            node->left = build(left, idx - 1);
            return node;
        };
        return build(0, n - 1);
    }
};
